test/test_linalg.py

```diff
@@ -914,6 +915,126 @@ def test_nuclear_norm_exceptions_old(self, device):
+            self.assertEqual(x_exp, x_act)  # Equality check
+            # TODO(@ivanyashchuk): remove this once batched matmul is avaiable on CUDA for complex dtypes
+            if self.device_type == 'cuda' and dtype.is_complex:
+                Ax = torch.matmul(A.cpu(), x_act.cpu()).to(device)
+            else:
+                Ax = torch.matmul(A, x_act)
+            self.assertEqual(b, Ax)
+
+        for batchsize in [1, 3, 4]:
+            solve_batch_helper((5, batchsize), (batchsize, 5, 10))
+
+    @skipCUDAIfNoMagma
+    @skipCPUIfNoLapack
+    @dtypes(torch.float32, torch.float64, torch.complex64, torch.complex128)
+    def test_solve_batched_non_contiguous(self, device, dtype):
+        from numpy.linalg import solve
+        from torch.testing._internal.common_utils import random_fullrank_matrix_distinct_singular_value
+        A = random_fullrank_matrix_distinct_singular_value(2, 2, dtype=dtype).to(device).permute(1, 0, 2)
+        b = torch.randn(2, 2, 2, dtype=dtype, device=device).permute(2, 1, 0)
+        x, _ = torch.solve(b, A)
+        x_exp = solve(A.cpu().numpy(), b.cpu().numpy())
+        self.assertEqual(x, x_exp)
+
+    @slowTest
+    @skipCUDAIfNoMagma
+    @skipCPUIfNoLapack
+    @dtypes(torch.float32, torch.float64, torch.complex64, torch.complex128)
+    def test_solve_batched_many_batches(self, device, dtype):
+        for A_dims, b_dims in zip([(5, 256, 256), (3, )], [(5, 1), (512, 512, 3, 1)]):
+            b, A = self.solve_test_helper(A_dims, b_dims, device, dtype)
+            x, _ = torch.solve(b, A)
+            # TODO(@ivanyashchuk): remove this once batched matmul is avaiable on CUDA for complex dtypes
+            if self.device_type == 'cuda' and dtype.is_complex:
+                Ax = torch.matmul(A.cpu(), x.cpu()).to(device)
+            else:
+                Ax = torch.matmul(A, x)
+            self.assertEqual(Ax, b.expand_as(x))
+
+    @skipCUDAIfNoMagma
+    @skipCPUIfNoLapack
+    @dtypes(torch.float32, torch.float64, torch.complex64, torch.complex128)
+    def test_solve_batched_broadcasting(self, device, dtype):
+        from numpy.linalg import solve
+
+        def run_test(A_dims, b_dims):
+            A_matrix_size = A_dims[-1]
+            A_batch_dims = A_dims[:-2]
+            b, A = self.solve_test_helper((A_matrix_size,) + A_batch_dims, b_dims, device, dtype)
+            x, _ = torch.solve(b, A)
+            x_exp = solve(A.cpu().numpy(), b.cpu().numpy())
+            self.assertEqual(x, x_exp)
+
+        # test against numpy.linalg.solve
+        run_test((2, 1, 3, 4, 4), (2, 1, 3, 4, 6))  # no broadcasting
+        run_test((2, 1, 3, 4, 4), (4, 6))  # broadcasting b
+        run_test((4, 4), (2, 1, 3, 4, 2))  # broadcasting A
+        run_test((1, 3, 1, 4, 4), (2, 1, 3, 4, 5))  # broadcasting A & b
+
+    @skipCUDAIfNoMagma
+    @skipCPUIfNoLapack
+    @dtypes(torch.float64, torch.complex128)
+    @dtypesIfCUDA(torch.float64)
+    def test_solve_autograd(self, device, dtype):
```

Comment:
I don't think we should add autograd tests in `test_linalg.py`. The only autograd tests this file are in `test_autograd_and_jit` which are tagged with a note that they should be moved to `method_tests` in `common_methods_invocations.py` cc. @mruberry

Reply:
for now we should just add tests in method_tests for solve by adding an entry here: https://github.com/pytorch/pytorch/blob/master/test/test_autograd.py#L4928-L4929 and adding complex specific tests similar to this (make sure to include 'complex' in the substring for each of the new tests so that they only run for complex dtype).

Reply:
It might also be worth adding LinalgFuncInfo to migrate the linalg autograd tests as we are extending the linalg support but it's ok to leave that for a follow-up PR.

Reply:
I recall @mruberry mentioned in other PRs to move all linalg related tests to `test_linalg.py`, so I think it includes also autograd tests that are not part of `method_tests`, and not only from `test_torch.py`.
Since in the future, the plan is to have LinalgFuncInfo instead of `method_tests`, I suggest we leave this test here and have CPU&CUDA variant tested without adding an entry to `method_tests` (which is a bit troublesome because we need to create an input with the dtype and device using `random_fullrank_matrix_distinct_singular_value(..., dtype, device)`.

Reply:
We often debate amongst ourselves where tests related to autograd should live. Do they live with the operator, or do they live with test_autograd.py? And we haven't come to a consistent resolution yet, unfortunately, so it's reviewers' discretion. In the (near) future I expect we'll publicize a definitive plan for where these tests go. Sorry there's not a clear solution at the moment.